File: src/characters/undertaker.ml
open! Core
open! Import

include Character.Make (struct
    type t =
      { kind : Kind.t
      ; alignment : Alignment.t
      }

    let id = "undertaker"
    let name = "Undertaker"
    let t = { kind = Kind.Townsfolk; alignment = Alignment.Good }
    let kind { kind; _ } = kind
    let alignment { alignment; _ } = alignment
  end)

let night_action ~player_id:pid ~night =
  if night = 1
  then None
  else
    Some
      (if_alive
         pid
         (let%bind.Botc_exec state = Botc_exec.get_state in
          match Game_state.last_execution state with
          | None -> Botc_exec.return ()
          | Some exec_id ->
            let%bind.Botc_exec () = Botc_exec.wake pid in
            let%bind.Botc_exec () =
              if Game_state.is_poisoned state pid
              then (
                let all_chars =
                  List.map (Game_state.seated_players state) ~f:(fun p ->
                    Player.character p)
                in
                let%bind.Botc_exec char = Botc_exec.narrator_pick all_chars in
                Botc_exec.tell pid (Char_display.name char))
              else (
                let exec_char =
                  Player.character (Map.find_exn (Game_state.players state) exec_id)
                in
                Botc_exec.tell pid (Char_display.name exec_char))
            in
            Botc_exec.sleep pid))
;;

let day_action ~player_id:_ = None
let on_setup ~player_id:_ = None
let on_nominated ~player_id:_ ~nominator:_ = None
let on_executed ~player_id:_ = None
let on_night_kill ~player_id:_ = None

File: src/characters/undertaker.rs
//! Undertaker (Townsfolk): each night after the first, learn which character
//! was executed earlier today.

use crate::ability::Ability;
use crate::ctx::Ctx;
use crate::ids::PlayerId;
use crate::role::{Alignment, CharacterInfo, Kind};
use crate::token::Token;

character!(
    Undertaker,
    "undertaker",
    "Undertaker",
    Kind::Townsfolk,
    Alignment::Good
);

impl Ability for Undertaker {
    fn info(&self) -> CharacterInfo {
        Self::INFO
    }

    fn on_night(&self, ctx: &mut Ctx, me: PlayerId, night: u32) {
        if night <= 1 || !ctx.grim.is_alive(me) {
            return;
        }
        // The player executed today carries the DiedToday token.
        let Some(dead) = ctx
            .grim
            .seats()
            .find(|&id| ctx.grim.has_token(id, Token::DiedToday))
        else {
            return; // no execution today: the Undertaker is not woken.
        };
        ctx.wake(me);
        let shown = if ctx.is_impaired(me) {
            let names = ctx.roster_character_names();
            names[ctx.choose("undertaker false info", &names)].clone()
        } else {
            let cid = ctx.registers_as_character(dead, "undertaker");
            ctx.registry().info(cid).name.to_string()
        };
        ctx.reveal(me, &format!("The {shown} was executed today"));
        ctx.sleep(me);
    }
}
